src/celeritas/grid/SplineXsCalculator.hh

```diff
@@ -0,0 +1,255 @@
+                                                     size_type low_idx,
+                                                     size_type high_idx) const;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from cross section data.
+ */
+CELER_FUNCTION
+SplineXsCalculator::SplineXsCalculator(XsGridData const& grid,
+                                       Values const& values,
+                                       size_type const order)
+    : data_(grid), reals_(values), loge_grid_(data_.log_energy), order_(order)
+{
+    CELER_EXPECT(data_);
+    CELER_ASSERT(grid.value.size() == data_.log_energy.size);
+
+    // order of interpolation must be smaller than the grid size for effective
+    // spline interpolation. Max order is set to be clipped at minimum and
+    // maximum energy indices so this may be unnecessary
+    CELER_ASSERT(order < grid.value.size());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the cross section.
+ *
+ * If needed, we can add a "log(energy/MeV)" accessor if we constantly reuse
+ * that value and don't want to repeat the `std::log` operation.
+ */
+CELER_FUNCTION real_type SplineXsCalculator::operator()(Energy energy) const
+{
+    real_type const loge = std::log(energy.value());
+
+    auto calc_extrapolated = [this, &energy](size_type idx) {
+        real_type result = this->get(idx);
+        if (idx >= data_.prime_index)
+        {
+            result /= energy.value();
+        }
+        return result;
+    };
+
+    // Snap out-of-bounds values to closest grid points
+    if (loge <= loge_grid_.front())
+    {
+        return calc_extrapolated(0);
+    }
+    if (loge >= loge_grid_.back())
+    {
+        return calc_extrapolated(loge_grid_.size() - 1);
+    }
+
+    // Locate the energy bin
+    size_type lower_idx = loge_grid_.find(loge);
+    CELER_ASSERT(lower_idx + 1 < loge_grid_.size());
+
+    // number of grid indexs away from the specified energy that need to be
+    // checked in both directions
+    size_type order_steps = order_ / 2 + 1;
+
+    // True bounding indices of the grid that will be checked.
+    // If the interpolation requests out-of-bounds indices, clip the
+    // extents. This will reduce the order of the interpolation
+    // todo - instead of clipping the bounds, alter both the low and high
+    // index to keep the range just shifted down
+
+    size_type true_low_idx;
+    if (lower_idx >= order_steps - 1)
+    {
+        true_low_idx = lower_idx - order_steps + 1;
+    }
+    else
+    {
+        true_low_idx = 0;
+    }
+    size_type true_high_idx
+        = min(lower_idx + order_steps + 1, loge_grid_.size());
+
+    // if the requested interpolation order is even, a direction must be
+    // selected to interpolate to
+    if (order_ % 2 == 0)
+    {
```

Comment:
It often improves readability to put the comment inside the condition it applies to:
```suggestion
    if (order_ % 2 == 0)
    {
        // The requested interpolation order is even: select a direction to interpolate toward
```

Reply:
Fixed